### PR TITLE
Dynamic aggregated lists

### PR DESCRIPTION
**Pull request by olemp**
### Your checklist for this pull request

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your main!
- [ ] Make sure you are making a pull request against the **dev** branch (left side). Also you should start *your branch* off *dev*.
- [ ] Check the commit's or even all commits' message 
- [ ] Check if your code additions will fail linting checks
- [ ] Remember: Add issue description to [CHANGELOG](https://github.com/Puzzlepart/prosjektportalen365/blob/dev/CHANGELOG.md) with the **ID of the issue** associated with this PR
- [ ] Documentation: Have a look at the [PP365 User manual](https://puzzlepart.github.io/prosjektportalen-manual/) and consider the need for updates to be made. Updates can be done directly into the 'Kladd' branch or by providing information to test team for implementation.

### Description

Please describe your pull request in detail.

### How to test

Please describe how someone else (a regular user without coding skills) can test this PR. In the form of a numbered list and checkboxes for Jan when the testing period occurs.

Example:

1. Do this
2. Then this
3. Then that

### Update of documentation
- [ ] Update user manual if needed

### CODE REVIEW

**Comment by Remi749** [comment]
@olemp, here is my question as discussed.

When trying to populate the filterpanel with state.columns. I get the following error: "Cannot add property selected, object is not extensible reduce redux state"

So it's trying to add the property 'selected' and set it to true when selecting a filter form the filterpanel. I've tried using current(state).columns but no luck.

As I'm new to redux I'm uncertain how to make this work. I just have to use the state.columns from the reducer the correct way.

**Comment by olemp** [comment]
> @olemp, here is my question as discussed.
> 
> When trying to populate the filterpanel with state.columns. I get the following error: "Cannot add property selected, object is not extensible reduce redux state"
> 
> So it's trying to add the property 'selected' and set it to true when selecting a filter form the filterpanel. This doesn't work as for some reason the state.columns is not mutable I guess. I've tried using current(state).columns but no luck.
> 
> As I'm new to redux I'm uncertain how to make this work. I just have to use the state.columns from the reducer the correct way.

The state is immutable, so you can't modify it. It's the best practice for state handling in React.

https://redux.js.org/faq/immutable-data

**Comment by Remi749** [comment]
> 
> The state is immutable, so you can't modify it. It's the best practice for state handling in React.
> 
> https://redux.js.org/faq/immutable-data

Okay,
Correction state.columns +> state.filters. 

Right now I'm adding columns to the filterPanel manually. changing the selected prop for this works as this is not immutable.

Populating the filterpanel with state.filters and then selecting/deselecting does not work. How can I make my new state.filters mutable?

**Comment by olemp** [comment]
> > The state is immutable, so you can't modify it. It's the best practice for state handling in React.
> > https://redux.js.org/faq/immutable-data
> 
> Okay, Correction state.columns -> state.filters (new state to contain all filters).
> 
> Right now I'm adding columns to the filterPanel manually. changing the selected prop for this works as this is not immutable.
> 
> Populating the filterpanel with state.filters and then selecting/deselecting does not work. How can I make my new state.filters mutable?

You don't want to make it mutable. You want to recreate it. 

https://stackoverflow.com/questions/34958775/why-should-objects-in-redux-be-immutable

File: SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx [hunk 1]
```diff
@@ -60,27 +62,33 @@ export const PortfolioAggregation = (props: IPortfolioAggregationProps) => {
...
-  const columns = useMemo(() => {
-    return filterItems(state.items, state.columns, state.activeFilters)
-  }, [state.columns, state.activeFilters])
+    return {
+      listItems: filteredItems.items.filter((i) => searchItem(i, state.searchTerm, state.columns)),
+      columns: filteredItems.columns
+    }
+  }, [state.searchTerm, state.items, state.activeFilters])
 
   const filters = useMemo(() => {
```
[the remark this answers is not in the document]
Comment (Remi749) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx:85]: Yes, that doesn't work. When the FilterPanel onChange tries to set the selected property I just get an error in regarding my previous issue with this. cloning the state.filters is the last thing that I tried that actually sorta works, but far from ideally and needs to be done correctly.

[the remark this answers is not in the document]
Comment (olemp) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx:85]: In the `onFilterChange`?

image

[the remark this answers is not in the document]
Comment (olemp) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: Hopefully https://github.com/Puzzlepart/prosjektportalen365/pull/707/commits/9e29d9b6cc16548ee1330e1598464e9434f1f38b does the trick.

[the remark this answers is not in the document]
Comment (Remi749) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: Now it works, but it works the same way as it did when I used the cloneDeep. When closing and opening the FilterPanel the state is not correct.

[the remark this answers is not in the document]
Comment (olemp) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: > the state is not correct.


What is incorrect about it? What is not *persisted*?

[the remark this answers is not in the document]
Comment (Remi749) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: image

^ Slik ser state filters ut etter at de i øverste bilde er huket bort. Ingen er satt til 'selected: false' som du ser

[the remark this answers is not in the document]
Comment (olemp) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: I don't think this has anything to do with Redux? There is something else fishy going on here I think?

[the remark this answers is not in the document]
Comment (olemp) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: The `state` seems to reset to `props` again when the panel is closed and opened again. The parent component (`PortfolioAggregation`) doesn't keep the `selected` property for `filters` as the `FilterPanel` does.

[the remark this answers is not in the document]
Comment (Remi749) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: I've tried to update the state.filters in the ON_FILTER_CHANGE action like so:

The first select/deselect is not registered to the state, but when selecting/deselecting another one, then the state is correct. Any ideas as to why? 

image

[the remark this answers is not in the document]
Comment (Remi749) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: First select/deselect:
image

Second select/deselct:
image

[the remark this answers is not in the document]
Comment (olemp) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: Maybe [this](https://github.com/Puzzlepart/prosjektportalen365/pull/707/commits/772113a77eb02b020d16b0a0fd413e1c6a16a46a) does the trick?

[the remark this answers is not in the document]
Comment (olemp) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: > First select/deselect: image Second select/deselct: image

Logging is from `PortfolioAggregation`? `FilterPanel` is modifying `filter` state - not the `PortfolioAggregation` component.

[the remark this answers is not in the document]
Comment (Remi749) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: > Maybe [this](https://github.com/Puzzlepart/prosjektportalen365/pull/707/commits/772113a77eb02b020d16b0a0fd413e1c6a16a46a) does the trick?

Now the FilterPanel is just empty 🤔

[the remark this answers is not in the document]
Comment (Remi749) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: Also Porteføljeoversikt FilterPanel doesn't save it's state. Probably due to latest commit?

[the remark this answers is not in the document]
Comment (olemp) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: @Remi749 I can't serve the solution now.

image

```
Error - [tsc] src/data/index.ts(144,32): error TS2345: Argument of type '"DATA_SOURCES"' is not assignable to parameter of type 'PortalDataServiceList'.
[0] Error - [tsc] src/data/index.ts(145,32): error TS2345: Argument of type '"PROJECT_CONTENT_COLUMNS"' is not assignable to parameter of type 'PortalDataServiceList'.
[0] Error - 'tsc' sub task errored after 7.7 s 
[0]  exited with code 2
```

[the remark this answers is not in the document]
Comment (Remi749) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: These should be available in the latest temp package of the shared npm package. It should be: 1.5.4-706.10

[the remark this answers is not in the document]
Comment (Remi749) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: Got it working! Committing soon

[the remark this answers is not in the document]
Comment (Remi749) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: It seems like activeFilters were never properly initialized. See https://github.com/Puzzlepart/prosjektportalen365/pull/707/commits/cb8c183f4801bc303f336b1c6c3b15a6ee720b17 for code additions

[the remark this answers is not in the document]
Comment (olemp) [reply SharePointFramework/PortfolioWebParts/src/components/PortfolioAggregation/index.tsx]: I did a fix that worked for `ON_FILTER_CHANGE` action, but got heaps of merge conflicts... You probably just need to redo the `fltColumns` stuff in the reducer @Remi749

**Pull request closed without merging**